### ISSUE TITLE
Fix inconsistent sanitization: sanitize on output rather than input

### ISSUE THREAD

**Issue by diosmosis** [matomo-org/matomo#4231]
The way in which Piwik deals w/ sanitizing user data is inconsistent: it is sometimes done on **input** (data is stored sanitized), sometimes on **output** (less frequently). For example, website names are stored in the DB sanitized and need to be unsanitized or outputted raw in HTML. Goal names are not sanitized in the DB and need to be escaped when outputting in HTML.

This should be made consistent. Additionally, [security best practices recommend](http://blog.ircmaxell.com/2011/03/what-is-security-web-application.html):
- filter on input (e.g. cast to an integer if you expect an integer, but don't sanitize strings)
- escape on **output**

Problems with the current approach:
- escaping on input is done for HTML/XML: that escaping doesn't makes sense for SQL, JavaScript/JSON, … And it needs to be undone for those cases
- double escaping issues if we escape on output (e.g. Twig and Angular do that by default)
  - #8496, #8123, #7987, #7969, #7806, #7531, #7528, #6821, #6722, #6325, #6068, #5189, #5009, #4749, #4709, #4231, #3954, #3549, #3503, #2519, #2400, #2386, #974, #341
- strings are stored in database escaped (not as their "real" values)

We should try to slowly move to sanitizing on output using native escaping features of Twig or AngularJS for example.

**Comment by mattab** [comment]
Note: there is another inconsistent sanitization policy in Piwik, see below function:

```
    /**
     * This function will sanitize or not if it's needed for the specified action type
     *
     * URLs (Page URLs, Downloads, Outlinks) are stored raw (unsanitized)
     * while other action types are stored Sanitized
     *
     * @param $actionType
     * @param $actionString
     * @return string
     */
    private static function normaliseActionString($actionType, $actionString)
```

and: `Action::isActionTypeStoredSanitized`

**Comment by mnapoli** [comment]
#6714 has been merged into this issue

**Comment by rogogon** [comment]
@mattab  hi, as this problem is open for a long time and commonly suffered, when are you planning to have it fixed?

**Comment by rogogon** [comment]
@mattab hi, as this problem is open for a long time and commonly suffered, when do you plan to fix it?

**Comment by mattab** [comment]
Due to the complexity of this task and the high risk of introducing security regressions (in particular, XSS), so far we have not been able to work / schedule this project. I would suspect it will take a long time before we can tackle this fully, and likely it will be done step by step over time, plugin by plugin etc. Any contribution will be very welcome

**Comment by mattab** [comment]
Thanks for creating this issue! We appreciate your input. This issue covers a lot of ground, and it's a bit too broad for us to tackle effectively as is. So, we're going to close it for now.

Instead some issues still opened are: https://github.com/matomo-org/matomo/issues/11786 https://github.com/matomo-org/matomo/issues/21164 https://github.com/matomo-org/matomo/issues/9223
